carcara/src/checker/rules/strings.rs

```diff
@@ -0,0 +1,136 @@
+    assert_clause_len, assert_eq, assert_num_args, assert_num_premises, get_premise_term, RuleArgs,
+    RuleResult,
+};
+use crate::ast::*;
+
+fn flatten(term: Rc<Term>, pool: &mut dyn TermPool) -> Vec<Rc<Term>> {
+    let mut flattened = Vec::new();
+    if let Term::Const(Constant::String(s)) = term.as_ref() {
+        flattened.extend(
+            s.chars()
+                .map(|c| pool.add(Term::Const(Constant::String(c.to_string())))),
+        );
+    } else if let Some(args) = match_term!((strconcat ...) = term) {
+        for arg in args {
+            flattened.extend(flatten(arg.clone(), pool));
+        }
+    } else {
+        flattened.push(term.clone());
+    }
+    flattened
+}
+
+fn concat_terms(terms: Vec<Rc<Term>>, pool: &mut dyn TermPool) -> Rc<Term> {
+    match terms.len() {
+        0 => pool.add(Term::Const(Constant::String(String::from("")))),
+        1 => terms[0].clone(),
+        _ => pool.add(Term::Op(Operator::StrConcat, terms)),
+    }
+}
+
+pub fn concat_eq(
+    RuleArgs {
+        premises,
+        args,
+        conclusion,
+        pool,
+        polyeq_time,
+        ..
+    }: RuleArgs,
+) -> RuleResult {
+    assert_num_premises(premises, 1)?;
+    assert_num_args(args, 1)?;
+    assert_clause_len(conclusion, 1)?;
+
+    let term = get_premise_term(&premises[0])?;
+    let rev = args[0].as_term()?.as_bool_err()?;
+    let (left, right) = match_term_err!((= l r) = term)?;
+    let mut left_flattened = flatten(left.clone(), pool);
+    let mut right_flattened = flatten(right.clone(), pool);
+
+    if rev {
+        left_flattened.reverse();
+        right_flattened.reverse();
+    }
+
+    let mut prefix = 0;
+    while (prefix < cmp::min(left_flattened.len(), right_flattened.len()))
+        && polyeq(
+            &left_flattened[prefix],
+            &right_flattened[prefix],
+            polyeq_time,
+        )
+    {
+        prefix += 1;
+    }
+
+    let mut left_suffix = left_flattened.get(prefix..).unwrap_or_default().to_vec();
+    let mut right_suffix = right_flattened.get(prefix..).unwrap_or_default().to_vec();
+
+    if rev {
+        left_suffix.reverse();
+        right_suffix.reverse();
+    }
+
+    let left_concat = concat_terms(left_suffix, pool);
+    let right_concat = concat_terms(right_suffix, pool);
+    let expected = pool.add(Term::Op(Operator::Equals, vec![left_concat, right_concat]));
+
+    let (conc_left, conc_right) = match_term_err!((= l r) = &conclusion[0])?;
+    let conc_left_flattened = flatten(conc_left.clone(), pool);
+    let conc_right_flattened = flatten(conc_right.clone(), pool);
+    let conc_left_concat = concat_terms(conc_left_flattened.clone(), pool);
+    let conc_right_concat = concat_terms(conc_right_flattened.clone(), pool);
+    let expanded_conc = pool.add(Term::Op(
+        Operator::Equals,
+        vec![conc_left_concat, conc_right_concat],
+    ));
+
+    assert_eq(&expected, &expanded_conc)
+}
+
+mod tests {
+    #[test]
+    fn concat_eq() {
```

Comment:
Overall the tests are pretty comprehensive, but could you add a few tests for invalid cases? For example, a test case where the prefixes don't match, or the argument given was wrong, etc. It's arguably more important to test for these cases because real-world proofs already test for the valid cases (and also, it's better to reject a valid proof than to accept an invalid one).

Reply:
Sure, I'll do that!